SBD management improve

liangxin1300 opened the pull request:
#### Motivation
We are facing several issues about sbd in crmsh lately.
It's might not a good way to add more codes without structure, and might cause more issues with small piece of code here and there.
It's better to manage sbd as a whole part.
#### Solution
Create a class `SBDManager`, and a set of methods to manage sbd configuration and services
#### Advantages
Treats sbd as a whole part
Specific methods are easy to find, understand and test
#### Changes

- Create class `SBDManager`
- `SBDManager.sbd_init` to get sbd device, initialize and write configure file
- `SBDManager.manage_sbd_service` to enable or disable sbd service
- `SBDManager.configure_sbd_resource` to configure stonith-sbd resource and stonith-enabled property
- `SBDManager.join_sbd` to check and verify sbd device on join process
- Make `-s`(for sbd device) option and `-S`(for diskless sbd) option mutual
- Unit test for `SBDManager`
- Functional test `bootstrap_sbd.feature` which can only running on local environment(since it seems impossible to configure watchdog inside container)

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,179 @@ def init_qdevice(self):
+            "SBD_DELAY_START": "no"
+            }
+
+    def __init__(self, sbd_device=None, diskless_sbd=False):
```
Maybe better to call `sbd_devices`, as the value can be a list

liangxin1300 replied:
Changed

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,179 @@ def init_qdevice(self):
+
+    def _get_sbd_device(self):
+        if self.sbd_device:
+            self.sbd_device = self._parse_sbd_device()
```
I don't know if it's a good idea to change the value of `sbd_device`, it might create issues. We should maybe create other parameter to store the values after parsing. What do you think?

liangxin1300 replied:
I'm not very clear here @arbulu89 , could you give me an example? Thanks!

liangxin1300 replied:
Trying to change this piece of code like below:
```
def _get_sbd_device(self):
        """
        Get sbd device from options or interactive mode
        """
        dev_list = []
        if self.sbd_devices:
            dev_list = self._parse_sbd_device()
            self._verify_sbd_device(dev_list)
        elif not self.diskless_sbd:
            dev_list = self._get_sbd_device_interactive()
        self.sbd_devices = dev_list
```

arbulu89 replied:
What I don't like is that the value of `self.sbd_devices` changes in format during the usage of this instance. You might have `sbd1;sbd2;sbd3` at the beginning and suddenly you have `['sbd', 'sbd2', 'sbd3']`. So it's a bit confusing. Usually doing this changes it's a bit risky, as you use the variable different in each situation.
That's why I suggested to have 2 variables like `self.sbd_devices_input` where you have always what the user pases, and `self._sbd_devices` after the transformation.
What do you think?

liangxin1300 replied:
OK, changed as `self.sbd_devices_input` and `self._sbd_devices`

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,179 @@ def init_qdevice(self):
+            warn("Not configuring SBD (%s left untouched)." % (SYSCONFIG_SBD))
+            return
+
+        status("""
```
Put this string in a constant for this Class, it will be better

liangxin1300 replied:
Changed

liangxin1300 replied:
Changed

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,179 @@ def init_qdevice(self):
+
+        dev_looks_sane = False
+        while not dev_looks_sane:
+            dev = prompt_for_string('Path to storage device (e.g. /dev/disk/by-id/...), or "none", use ";" as separator for multi path', r'none|\/.*')
```
Maybe we should add an improved message like `... or "none" for diskless sbd, use ";" ...` . What do you think?

liangxin1300 replied:
OK, Changed

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,179 @@ def init_qdevice(self):
+            if dev == "none":
+                self.diskless_sbd = True
+                return
+            dev_list = dev.strip(';').split(';')
```
I think we can somehow use `_verify_sbd_device` method here, the code is exactly the same. Not just copy/pasting, what with some minor changes it might be done

liangxin1300 replied:
Changed, reuse `_verify_sbd_device`

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -1375,136 +1536,14 @@ def sysconfig_comment_out(scfile, key):
-    status_done()
-
-
 def init_sbd():
```
I think we can remove this method, right?

liangxin1300 replied:
Currently, we can't do that. `init_xxx` was a format for `stage`, used by `globals()["init_" + stage]()`

arbulu89 replied:
Ok, I see. Thanks for the clarification

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,179 @@ def init_qdevice(self):
+    """
+    Class SBDManager is for sbd configuration and sbd service management
+    """
+    sbd_config_dict = {
```
This dict doesn't look like follows PEP8 indentation rules

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -2016,10 +2047,8 @@ def update_nodeid(nodeid, node=None):
-    # check the existing cluster if the sbd service is enabled
-    if not configured_sbd_device() and invoke("ssh -o StrictHostKeyChecking=no root@{} systemctl is-enabled sbd.service".format(seed_host)):
-        _context.diskless_sbd = True
+    if invoke("ssh -o StrictHostKeyChecking=no root@{} systemctl is-enabled sbd.service".format(seed_host)):
```
Why don't you put this piece of code in the `SBDManager`? It totally belongs to sbd

liangxin1300 replied:
I put this into `detect_remote_sbd_service_status` method

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,179 @@ def init_qdevice(self):
+    """
+    Class SBDManager is for sbd configuration and sbd service management
+    """
+    sbd_config_dict = {
```
This variable is not so properly used. You declare here as constant or class method, but you change it for each instance. That's not a good idea.  If you are using for a particular object instance class (what you do right now), just declare this method like the others, and not as class method

liangxin1300 replied:
@arbulu89 , I put `sbd_config_dict` into function `_update_configuration` as a internal variables, what do you think of this change?

arbulu89 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,179 @@ def init_qdevice(self):
+        # If process work through here, consider it's ready for enable service
+        self._sbd_service_flag = True
+
+    def configured_sbd_device(self):
```
I would change the name of this method to `get_sbd_configuration`

liangxin1300 replied:
Changed

liangxin1300 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,179 @@ def init_qdevice(self):
+
+class SBDManager(object):
+    """
+    Class SBDManager is for sbd configuration and sbd service management
```
Changed

liangxin1300 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,179 @@ def init_qdevice(self):
+
+    def _check_environment(self):
+        if not check_watchdog():
+            error("Watchdog device must be configured if want to use SBD!")
```
Changed

liangxin1300 commented:
@arbulu89 Thanks a lot! Your suggestions are really helpful for me!

> 
> * Add a docstring to the methods please (to all of them). It's not a big effort but makes easier to understand the purpose of the method to newcomers (like me hehe)

Changed

> * I think the methods inside `SBDManager` are strangely ordered. Put in the top methods like `_check_environment` and `_parse_sbd_device` as they are used by `_sbd_init` (I would put them before that method). Otherwise we need to be scrolling up and down all the time.

I've adjusted the order of methods.

> * I have seen many methods that don't use anything of the instance (`self` parameters). I would convert those methods to `Classmethod` or `staticmethod`. It makes sense

I've converted those methods to `staticmethod`.
Thank you!

> 
> PD: With these changes the code looks much cleaner. Eventually, we should do the same for all the methods in the init and join processes, to have a neat code

liangxin1300 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,201 @@ def init_qdevice(self):
+    """
+    Class to manage sbd configuration and services
+    """
+    sbd_status_description = """
```
Changed

liangxin1300 commented on crmsh/bootstrap.py:
```diff
@@ -104,6 +105,201 @@ def init_qdevice(self):
+                self.diskless_sbd = True
+                return
+            dev_list = dev.strip(';').split(';')
+            self._verify_sbd_device(dev_list)
```
Changed as it can continue again with an error message

gao-yan commented on crmsh/bootstrap.py:
```diff
@@ -105,6 +106,225 @@ def init_qdevice(self):
+        status(self.SBD_STATUS_DESCRIPTION)
+
+        if not confirm("Do you wish to use SBD?"):
+            warn("Not configuring SBD - STONITH will be disabled.")
```
I don't think we are going to turn off `stonith-enabled` cluster property in this case, right? To avoid confusions, we could  probably say something like  "Not configuring SBD - Fencing won't work without fencing resources being configured ".

gao-yan commented on crmsh/bootstrap.py:
```diff
@@ -105,6 +106,225 @@ def init_qdevice(self):
+        On joining process, check whether peer node has enabled sbd.service
+        If so, check prerequisites of SBD and verify sbd device on join node
+        """
+        if not os.path.exists(SYSCONFIG_SBD):
```
Just wondering,  does it mean /etc/sysconfig/sbd is supposed to have already been synchronized to this joining node if there's any on the first node?
